delete dm endpoints

majestrate opened the pull request:
fixes #151

majestrate committed:
add endpoint to delete own DMs

added DELETE /inbox endpoint that deletes all direct messages they
recieved.

jagerman commented on setup.cfg:
```diff
@@ -21,3 +21,5 @@ install_requires=
     oxenmq
     oxenc
     pyonionreq
+    tomil
```
What is this?  I can't find any such python package.

majestrate replied:
typo, it is tomli

jagerman commented on sogs/http.py:
```diff
@@ -1,6 +1,7 @@
 # success codes:
 OK = 200
 CREATED = 201
+NO_CONTENT = 204
```
I don't think we should use this; everywhere else in SOGS where we have nothing to return we're still returning application/json content of an empty dict (`{}`) with 200 OK, which allows us flexibility to add return values later without changing the status code.

jagerman commented on sogs/routes/auth.py:
```diff
@@ -121,6 +121,27 @@ def required_user_wrapper(*args, **kwargs):
+    Unauthorized if the user has not authenticated with a blinded pubkey."""
+    require_user()
+    if not g.user.is_blinded:
+        abort_with_reason(http.FORBINDDEN, "This endpoint requires blinded pubkeys be used")
```
Typo: `FORBIDDEN`

jagerman commented on sogs/model/message.py:
```diff
@@ -42,6 +42,25 @@ def __init__(self, row=None, *, sender=None, recip=None, data=None):
+        """Delete all messages sent to a user or from a user.
+        returns the number of rows affected.
+        """
+        assert recip or sender
```
This should throw an exception rather than a debug assert.  But also, given the query construction below, it needs to be the case that the caller has specified exactly one of these, rather than at least one: otherwise this function is going to generate an invalid query.

jagerman commented on sogs/routes/dm.py:
```diff
@@ -112,3 +110,25 @@ def send_inbox(sid):
+    num_deleted = 0
+    with db.transaction():
+        num_deleted += Message.delete_all(recip=g.user)
+        num_deleted += Message.delete_all(sender=g.user)
```
I'm don't think we should allow deleting sent messages.  Sent messages are the property of the recipient in regular Session message model (at best you can send another deletion meta-request follow-up) and I'm thinking that should be mirrored here.

jagerman replied:
Additionally, this request is for `DELETE /inbox`, so deleting everything in the outbox as well doesn't seem right.

(But as to my previous point: I don't think we should have a `DELETE /outbox`).

jagerman commented on sogs/routes/dm.py:
```diff
@@ -112,3 +110,25 @@ def send_inbox(sid):
+    # Return value
+
+    204 No Content - On successful removal of messages.
+    304 Not Modified - If there were no messages to remove.
```
I don't like either of these statuses:

- On success everywhere else in SOGS we return a 200 OK with a json result (even if empty).

- 304 is a redirection response code for cache-control headers which which seems deeply inappropriate for a DELETE request response.  A 200 response code seems appropriate in this case: we asked to delete all, and it succeeded; the fact that there were none to delete doesn't mean it didn't have the intended effect.

jagerman commented on sogs/routes/dm.py:
```diff
@@ -112,3 +110,25 @@ def send_inbox(sid):
+
+    if num_deleted:
+        return b'', http.NO_CONTENT
+    return b'', http.NOT_MODIFIED
```
Since we have it, we might as well include the number of deleted messages in the response.

jagerman commented on sogs/model/message.py:
```diff
@@ -42,6 +42,25 @@ def __init__(self, row=None, *, sender=None, recip=None, data=None):
+            DELETE FROM inbox WHERE
+            {'recipient = :recip' if recip else '' }
+            {'sender = :sender' if sender else '' }
+            """,
```
This query is going to be invalid if you pass in both a `recip` and ` sender` at the same time.

jagerman commented on sogs/routes/auth.py:
```diff
@@ -121,6 +121,27 @@ def required_user_wrapper(*args, **kwargs):
 
+def require_blind_user():
+    """Requires that the authenticated user is using a blinded pubkey for auth; aborts with 401
+    Unauthorized if the user has not authenticated with a blinded pubkey."""
```
Status code in the docstring doesn't match the code.

jagerman commented on sogs/model/message.py:
```diff
@@ -42,6 +42,27 @@ def __init__(self, row=None, *, sender=None, recip=None, data=None):
+        returns the number of rows affected.
+        """
+        if sender and recip:
+            raise ValueError("delete_all(): cannot provide both sender and recipient")
```
Also needs to fail if you pass in neither (because otherwise this query is going to delete all DMs of all users if two `None`s get passed in here).

jagerman committed:
Require either sender or recipient for deletion

Otherwise passing in None (or calling without arguments) deletes all
messages for all users, which is unexpected/undesirable.

jagerman committed:
docstring formatting/rewording
